graders/python/python_autograder/pl_execute.py

```diff
@@ -96,37 +106,45 @@ def execute_code(
         os.remove(path.join(filenames_dir, "trailing_code.py"))
     os.remove(path.join(filenames_dir, "test.py"))
 
-    repeated_setup_name = "repeated_setup()"
-    if repeated_setup_name not in str_setup:
-        repeated_setup_name = "pass"
+    # Since we've deleted some files, we need to manually populate
+    # the linecache so that `traceback` can find the correct contents when
+    # printing any exceptions.
+    populate_linecache(path.join(filenames_dir, "setup_code.py"), str_setup)
+    populate_linecache(fname_ref, str_ref)
 
     # Seed student code and answer code with same seed
     seed = random.randint(0, (2**32) - 1)
 
-    setup_code = {"test_iter_num": test_iter_num, "data": data}
+    setup_globals = {"test_iter_num": test_iter_num, "data": data}
     # make all the variables in setup_code.py available to ans.py
-    exec(str_setup, setup_code)
-    exec(repeated_setup_name, setup_code)
+    code_setup = compile(str_setup, path.join(filenames_dir, "setup_code.py"), "exec")
+    exec(code_setup, setup_globals)
+
+    # If the setup code has a repeated_setup function, run it.
+    repeated_setup = setup_globals.get("repeated_setup")
+    if repeated_setup is not None and callable(repeated_setup):
+        repeated_setup()
```

Comment:
This is much more direct and reliable than checking for the presence of the string `repeated_setup()` in the course code. For instance, this allows for a commented-out `repeated_setup()` function (which is something I actually hit while testing all this).

Reply:
Agreed, that old code is... interesting.